File: zh-hant/codes/rust/chapter_greedy/max_capacity.rs
/*
 * File: coin_change_greedy.rs
 * Created Time: 2023-07-22
 * Author: night-cruise (2586447362@qq.com)
 */

/* 最大容量：貪婪 */
fn max_capacity(ht: &[i32]) -> i32 {
    // 初始化 i, j，使其分列陣列兩端
    let mut i = 0;
    let mut j = ht.len() - 1;
    // 初始最大容量為 0
    let mut res = 0;
    // 迴圈貪婪選擇，直至兩板相遇
    while i < j {
        // 更新最大容量
        let cap = std::cmp::min(ht[i], ht[j]) * (j - i) as i32;
        res = std::cmp::max(res, cap);
        // 向內移動短板
        if ht[i] < ht[j] {
            i += 1;
        } else {
            j -= 1;
        }
    }
    res
}

/* Driver Code */
fn main() {
    let ht = [3, 8, 5, 2, 7, 7, 3, 4];

    // 貪婪演算法
    let res = max_capacity(&ht);
    println!("最大容量為 {}", res);
}

File: codes/lua/chapter_greedy/max_capacity.lua
-- @script max_capacity.lua
-- @date 2025-11-16
-- @author fisheryv (yue.fisher2025@gdhfi.com)

--- 最大容量：贪心
--- @param ht table 高度数组
--- @return integer 最大容量
local function max_capacity(ht)
    -- 初始化 i, j，使其分列数组两端
    local i = 1
    local j = #ht
    -- 初始最大容量为 0
    local res = 0
    -- 循环贪心选择，直至两板相遇
    while i < j do
        -- 更新最大容量
        local cap = math.min(ht[i], ht[j]) * (j - i)
        res = math.max(res, cap)
        -- 向内移动短板
        if ht[i] < ht[j] then
            i = i + 1
        else
            j = j - 1
        end
    end
    return res
end

-- Driver Code
local function main()
    local ht = { 3, 8, 5, 2, 7, 7, 3, 4 }

    -- 贪心算法
    local res = max_capacity(ht)
    print(string.format("最大容量为 %d", res))
end

-- 执行主函数
main()
